### PR TITLE
feat: testing transactions and testing addresses

### PR DESCRIPTION
**Pull request by ubavic**

### CODE REVIEW

File: lib/cash-in/cash-in-tx.js [hunk 1]
```diff
@@ -94,6 +98,14 @@ function checkForBlacklisted (tx) {
   return Promise.resolve(false)
 }
 
+function checkForTestingAddress (tx) {
+  // Check only on addressScan and avoid testing for blacklist on every bill inserted
```
Comment (chaotixkilla) [inline lib/cash-in/cash-in-tx.js]: Forgot to change the comment? 😄

File: migrations/1662381432402-test-transactions.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+exports.up = function (next) {
+  const sql = [
+    `CREATE TABLE testing_addresses (
+        crypto_code TEXT NOT NULL,
+        address TEXT NOT NULL,
+        unique (crypto_code, address)
+      )`,
+    `ALTER TABLE cash_out_txs ADD COLUMN testing BOOLEAN NOT NULL DEFAULT FALSE`,
+    `ALTER TABLE cash_in_txs ADD COLUMN testing BOOLEAN NOT NULL DEFAULT FALSE`,
```
Comment (chaotixkilla) [inline migrations/1662381432402-test-transactions.js]: Would prefer a more boolean-like name for this flag, since `testing` seems vague when reading the object as a whole. Something like `is_test_transaction` goes more directly straight to the meaning of this flag

File: new-lamassu-admin/src/pages/TestingAddresses/TestingAddresses.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    }
+    const res = await addEntry({ variables: { cryptoCode, address } })
+    if (!res.errors) {
+      return setShowModal(false)
+    }
+    const duplicateKeyError = res.errors.some(e => {
+      return e.message.includes('duplicate')
+    })
+    if (duplicateKeyError) {
+      setErrorMsg('This address is already on list')
```
Comment (chaotixkilla) [inline new-lamassu-admin/src/pages/TestingAddresses/TestingAddresses.js]: A bit nitpicky, but 'on list' sounds weird. Could we consider `This address is already listed` or `This address is already on the list`?

File: new-lamassu-admin/src/pages/TestingAddresses/TestingAddressesModal.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    BTC: '1ADwinnimZKGgQ3dpyfoUZvJh4p1UWSSpD',
+    ETH: '0x71C7656EC7ab88b098defB751B7401B5f6d8976F',
+    LTC: 'LPKvbjwV1Kaksktzkr7TMK3FQtQEEe6Wqa',
+    DASH: 'XqQ7gU8eM76rEfey726cJpT2RGKyJyBrcn',
+    ZEC: 't1KGyyv24eL354C9gjveBGEe8Xz9UoPKvHR',
+    BCH: 'qrd6za97wm03lfyg82w0c9vqgc727rhemg5yd9k3dm',
+    USDT: '0x5754284f345afc66a98fbb0a0afe71e0f007b949',
+    XMR:
+      '888tNkZrPN6JsEgekjMnABU4TBzc2Dt29EPAvkRxbANsAnjyPbb3iQ1YBRk1UXcdRsiKc9dhwMVgN5S9cQUiyoogDavup3H'
+  }
```
Comment (chaotixkilla) [inline new-lamassu-admin/src/pages/TestingAddresses/TestingAddressesModal.js]: Since this is repeated from the blacklist page, could we just move the variable up to a constants file? Admin one should be in `src/utils/constants.js`

File: lib/cash-in/cash-in-tx.js [hunk 5]
```diff
@@ -35,12 +36,14 @@ function post (machineTx, pi) {
       let blacklisted = false
       let addressReuse = false
       let walletScore = {}
+      let testingAddress = false
 
-      return Promise.all([settingsLoader.loadLatest(), checkForBlacklisted(updatedTx), doesTxReuseAddress(updatedTx), getWalletScore(updatedTx, pi)])
-        .then(([{ config }, blacklistItems, isReusedAddress, fetchedWalletScore]) => {
+      return Promise.all([settingsLoader.loadLatest(), checkForBlacklisted(updatedTx), doesTxReuseAddress(updatedTx), getWalletScore(updatedTx, pi), checkForTestingAddress(updatedTx)])
+        .then(([{ config }, blacklistItems, isReusedAddress, fetchedWalletScore, isAddressTesting]) => {
```
Comment (chaotixkilla) [inline lib/cash-in/cash-in-tx.js]: Another nitpicky thing, but the variable name makes more sense this way imo

```suggestion
        .then(([{ config }, blacklistItems, isReusedAddress, fetchedWalletScore, isTestingAddress]) => {
```

File: lib/new-admin/graphql/resolvers/transaction.resolver.js [hunk 6]
```diff
@@ -36,7 +36,8 @@ const resolvers = {
   },
   Mutation: {
     cancelCashOutTransaction: (...[, { id }]) => cashOutTx.cancel(id),
-    cancelCashInTransaction: (...[, { id }]) => cashInTx.cancel(id)
+    cancelCashInTransaction: (...[, { id }]) => cashInTx.cancel(id),
+    setTesting: (...[, { txClass, id, testing }]) => transactions.setTesting(txClass, id, testing)
```
Comment (chaotixkilla) [inline lib/new-admin/graphql/resolvers/transaction.resolver.js]: Better to have a mutation for enabling an address as testing and another for disabling. This way we don't rely on a potentially faulty value coming from the front-end (check our testing customer flag on `customer.type.js`)

```suggestion
    enableTestTransaction: (...[, { txClass, id}]) => transactions.enableTestTransaction(txClass, id),
    disableTestTransaction: (...[, { txClass, id}]) => transactions.disableTestTransaction(txClass, id)
```

File: lib/new-admin/graphql/types/transaction.type.js [hunk 7]
```diff
@@ -69,6 +70,7 @@ const typeDef = gql`
   type Mutation {
     cancelCashOutTransaction(id: ID): Transaction @auth
     cancelCashInTransaction(id: ID): Transaction @auth
+    setTesting(txClass: String, id: ID, testing: Boolean): Transaction @auth
```
Comment (chaotixkilla) [inline lib/new-admin/graphql/types/transaction.type.js]: Same here

File: lib/new-admin/services/transactions.js [hunk 8]
```diff
@@ -362,12 +362,22 @@ function updateTxCustomerPhoto (customerId, txId, direction, data) {
...
 
+function setTesting (txClass, txID, testing) {
+  const sql = `
+    UPDATE ${txClass === 'cashIn' ? 'cash_in_txs' : 'cash_out_txs' } 
+    SET testing=$2
+    WHERE id=$1`
+
+  return db.oneOrNone(sql, [txID, testing])
+}
+
```
Comment (chaotixkilla) [inline lib/new-admin/services/transactions.js]:
```suggestion
function enableTestTransaction (txClass, txID) {
  const sql = `
    UPDATE ${txClass === 'cashIn' ? 'cash_in_txs' : 'cash_out_txs' } 
    SET testing = true
    WHERE id=$1`

  return db.oneOrNone(sql, [txID])
}

function disableTestTransaction (txClass, txID) {
  const sql = `
    UPDATE ${txClass === 'cashIn' ? 'cash_in_txs' : 'cash_out_txs' } 
    SET testing = false
    WHERE id=$1`

  return db.oneOrNone(sql, [txID])
}

```

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 9]
```diff
@@ -86,6 +87,14 @@ const CANCEL_CASH_IN_TRANSACTION = gql`
   }
 `
 
+const SET_TESTING = gql`
+  mutation AddTestingAddress($txClass: String!, $id: ID!, $testing: Boolean!) {
+    setTesting(txClass: $txClass, id: $id, testing: $testing) {
```
Comment (chaotixkilla) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: Same here

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 10]
```diff
@@ -119,6 +128,11 @@ const DetailsRow = ({ it: tx, timezone }) => {
     }
   )
 
+  const [setTesting] = useMutation(SET_TESTING, {
+    onError: ({ message }) => setErrorMessage(message ?? 'An error occurred.'),
+    refetchQueries: () => ['transactions']
+  })
```
Comment (chaotixkilla) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: This would become two mutations

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 11]
```diff
@@ -411,6 +425,23 @@ const DetailsRow = ({ it: tx, timezone }) => {
...
+            <Switch
+              checked={tx.testing}
+              onChange={() =>
+                setTesting({
+                  variables: {
+                    txClass: tx.txClass,
+                    id: tx.id,
+                    testing: !tx.testing
+                  }
+                })
```
Comment (chaotixkilla) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]:
```suggestion
              onChange={() =>
                R.path(['isTesting'])(tx)
                  ? disableTestTransaction({ variables: { txClass: tx.txClass, id: tx.id } })
                  : enableTestTransaction({ variables: { txClass: tx.txClass, id: tx.id } })
```

**Review by chaotixkilla — CHANGES_REQUESTED** [review]
